File: public/cpp/064-partition-equal-subset-sum.cpp
/*
 * Problem: Partition Equal Subset Sum (LeetCode 416)
 * Link: https://leetcode.com/problems/partition-equal-subset-sum/
 * Difficulty: Medium
 * Category: Dynamic Programming
 *
 * Description:
 * Given an integer array nums, return true if you can partition the array into
 * two subsets such that the sum of the elements in both subsets is equal or false
 * otherwise.
 *
 * Example 1:
 * Input: nums = [1,5,11,5]
 * Output: true
 * Explanation: [1,5,5] and [11]
 *
 * Example 2:
 * Input: nums = [1,2,3,5]
 * Output: false
 *
 * Constraints:
 * - 1 <= nums.length <= 200
 * - 1 <= nums[i] <= 100
 */

#include <vector>
#include <algorithm>
#include <iostream>
#include <numeric>
#include <bitset>
using namespace std;

/*
 * APPROACH 1: 2D DP (Boolean Table)
 *
 * Intuition:
 * - Total sum must be even; target = sum/2
 * - dp[i][j] = can we form sum j using first i items?
 * - Recurrence: dp[i][j] = dp[i-1][j] || dp[i-1][j-nums[i-1]]
 *
 * Time:  O(n * target)
 * Space: O(n * target)
 */
class Solution1 {
public:
    bool canPartition(vector<int>& nums) {
        int total = accumulate(nums.begin(), nums.end(), 0);
        if (total % 2 != 0) return false;
        int target = total / 2;
        int n = nums.size();
        vector<vector<bool>> dp(n + 1, vector<bool>(target + 1, false));
        for (int i = 0; i <= n; i++) dp[i][0] = true;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= target; j++) {
                dp[i][j] = dp[i-1][j];
                if (j >= nums[i-1])
                    dp[i][j] = dp[i][j] || dp[i-1][j - nums[i-1]];
            }
        }
        return dp[n][target];
    }
};

/*
 * APPROACH 2: 1D DP (Space Optimised)
 *
 * Intuition:
 * - Reduce 2D table to 1D by traversing j right-to-left
 * - dp[j] = can we form sum j with elements seen so far?
 *
 * Time:  O(n * target)
 * Space: O(target)
 */
class Solution2 {
public:
    bool canPartition(vector<int>& nums) {
        int total = accumulate(nums.begin(), nums.end(), 0);
        if (total % 2 != 0) return false;
        int target = total / 2;
        vector<bool> dp(target + 1, false);
        dp[0] = true;
        for (int num : nums) {
            for (int j = target; j >= num; j--) {
                dp[j] = dp[j] || dp[j - num];
            }
        }
        return dp[target];
    }
};

/*
 * APPROACH 3: Memoization (Top-Down)
 *
 * Intuition:
 * - Recursively try including / excluding each element
 * - Cache results in memo[i][rem]
 *
 * Time:  O(n * target)
 * Space: O(n * target)
 */
class Solution3 {
private:
    vector<vector<int>> memo;
    bool helper(vector<int>& nums, int i, int rem) {
        if (rem == 0) return true;
        if (i == 0 || rem < 0) return false;
        if (memo[i][rem] != -1) return memo[i][rem];
        memo[i][rem] = helper(nums, i-1, rem) || helper(nums, i-1, rem - nums[i-1]);
        return memo[i][rem];
    }
public:
    bool canPartition(vector<int>& nums) {
        int total = accumulate(nums.begin(), nums.end(), 0);
        if (total % 2 != 0) return false;
        int target = total / 2;
        memo.assign(nums.size() + 1, vector<int>(target + 1, -1));
        return helper(nums, nums.size(), target);
    }
};

/*
 * APPROACH 4: Bitset Optimisation
 *
 * Intuition:
 * - Represent reachable sums as bits in a bitset
 * - For each num, shift bits left by num and OR into the set
 * - Check if target bit is set
 *
 * Time:  O(n * target / 64) — bitset parallelism
 * Space: O(target)
 */
class Solution4 {
public:
    bool canPartition(vector<int>& nums) {
        int total = accumulate(nums.begin(), nums.end(), 0);
        if (total % 2 != 0) return false;
        int target = total / 2;
        bitset<10001> bits;
        bits[0] = 1;
        for (int num : nums) bits |= (bits << num);
        return bits[target];
    }
};

void test(vector<int> nums, int approach) {
    bool result;
    cout << "Input: [";
    for (int i = 0; i < (int)nums.size(); i++) { cout << nums[i]; if (i+1<(int)nums.size()) cout<<","; }
    cout << "] ";
    switch(approach) {
        case 1: { Solution1 s; result = s.canPartition(nums); cout << "2D-DP: "; break; }
        case 2: { Solution2 s; result = s.canPartition(nums); cout << "1D-DP: "; break; }
        case 3: { Solution3 s; result = s.canPartition(nums); cout << "Memo:  "; break; }
        case 4: { Solution4 s; result = s.canPartition(nums); cout << "Bitset:"; break; }
    }
    cout << (result ? "true" : "false") << "\n";
}

int main() {
    vector<int> t1 = {1,5,11,5};   for(int i=1;i<=4;i++) test(t1,i);
    vector<int> t2 = {1,2,3,5};    for(int i=1;i<=4;i++) test(t2,i);
    vector<int> t3 = {1,1};        for(int i=1;i<=4;i++) test(t3,i);
    vector<int> t4 = {3,3,3,4,5};  for(int i=1;i<=4;i++) test(t4,i);
    return 0;
}

/*
 * COMPLEXITY SUMMARY:
 * Approach 1 (2D DP):    Time O(n*T)        Space O(n*T)
 * Approach 2 (1D DP):    Time O(n*T)        Space O(T)
 * Approach 3 (Memo):     Time O(n*T)        Space O(n*T) + stack
 * Approach 4 (Bitset):   Time O(n*T/64)     Space O(T)
 *
 * T = target = sum/2
 */

// Made with Bob

File: src/data/tutorials/064-partition-equal-subset-sum.js
/**
 * Tutorial content for #064 — Partition Equal Subset Sum
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given an array of positive integers, determine if it can be partitioned into two subsets with equal sum. Each element must go into exactly one subset.`,
    example: `nums = [1, 5, 11, 5]\n→ Subset 1: [11], Subset 2: [1, 5, 5] both sum to 11\n✅ Answer: true`,
    keyInsight: `Equal partition means each subset sums to totalSum / 2. If totalSum is odd, it's impossible. So the problem reduces to: can we find a subset that sums to exactly totalSum / 2? This is the classic 0/1 knapsack.`,
  },

  approaches: {
    '2D Boolean DP': {
      intuition: `Build a 2D boolean table dp[i][j] = "can we form sum j using the first i elements?" Seed dp[i][0]=true (empty subset sums to 0). For each element, either include it (dp[i-1][j-nums[i-1]]) or exclude it (dp[i-1][j]).`,
      steps: [
        `Compute target = totalSum / 2; if totalSum is odd, return false.`,
        `Create dp[n+1][target+1]; set dp[i][0]=true for all i.`,
        `For i from 1 to n, for j from 1 to target:`,
        `  dp[i][j] = dp[i-1][j] (exclude nums[i-1]).`,
        `  If j >= nums[i-1]: dp[i][j] |= dp[i-1][j - nums[i-1]] (include).`,
        `Return dp[n][target].`,
      ],
      example: `nums=[1,5,11,5], sum=22, target=11\n\nAfter num=1:  can reach: {0,1}\nAfter num=5:  can reach: {0,1,5,6}\nAfter num=11: can reach: {0,1,5,6,11,12,16,17}\nAfter num=5:  can reach 11? Yes (5+1+5=11 or 6+5=11) ✓\n✅ Answer: true`,
      keyInsight: `O(n × target) time and space. Explicit 2D table is easy to trace but uses O(n × target) memory.`,
    },

    'Space-Optimized 1D DP': {
      intuition: `The 2D DP only looks back one row. Use a single boolean array dp[j] = "can we form sum j?". Iterate nums one by one. For each num, iterate j from target down to num (right-to-left prevents using the same element twice).`,
      steps: [
        `Compute target = totalSum / 2; if odd, return false.`,
        `Create dp[target+1] all false; dp[0] = true.`,
        `For each num in nums:`,
        `  For j from target down to num:`,
        `    dp[j] |= dp[j - num].`,
        `Return dp[target].`,
      ],
      example: `nums=[1,5,11,5], target=11\n\ndp=[T,F,F,F,F,F,F,F,F,F,F,F]\nnum=1:  dp[1]=dp[0]=T → dp=[T,T,F,...]\nnum=5:  dp[6]=dp[1]=T; dp[5]=dp[0]=T\nnum=11: dp[11]=dp[0]=T ✓ → return true immediately\n✅ Answer: true`,
      keyInsight: `O(n × target) time, O(target) space. The right-to-left inner loop is critical — it ensures each element is used at most once (0/1 knapsack constraint).`,
    },

    'Memoization (Top-Down)': {
      intuition: `Recursively decide for each element: include it (reduce the remaining target) or skip it. Cache each (index, remaining-sum) pair so repeated sub-problems are solved once.`,
      steps: [
        `Compute target = totalSum / 2; if odd, return false.`,
        `Create memo[n][target+1] filled with -1.`,
        `Base cases: remaining==0 → true; index>=n or remaining<0 → false.`,
        `If memo[index][remaining] != -1, return it.`,
        `memo[index][remaining] = solve(index+1, remaining-nums[index]) OR solve(index+1, remaining).`,
        `Return memo[0][target].`,
      ],
      example: `nums=[1,5,11,5], target=11\n\nsolve(0,11): include 1 → solve(1,10); OR skip → solve(1,11)\nsolve(1,10): include 5 → solve(2,5); OR skip → solve(2,10)\nsolve(2,5): include 11→ solve(3,-6) false; skip → solve(3,5)\nsolve(3,5): include 5 → solve(4,0) → true ✅\n✅ Answer: true`,
      keyInsight: `O(n × target) time and space. Top-down naturally prunes paths early when remaining < 0, which can be faster in practice than filling the whole table.`,
    },

    'Bitset Optimisation': {
      intuition: `Represent reachable sums as a bitset (big integer or long). Initially bits[0]=1. For each number, shift the bitset left by num (meaning "add num to all achievable sums") and OR it with the current bitset. Check if bit at target position is set.`,
      steps: [
        `Compute target = totalSum / 2; if odd, return false.`,
        `Use a BigInteger or long bits = 1 (bit 0 is set).`,
        `For each num: bits |= (bits << num).`,
        `Return bit at position target is set.`,
      ],
      example: `nums=[1,5,11,5], target=11\n\nbits = 1 (binary: ...0001)\nnum=1:  bits |= bits<<1 → ...0011 (can reach 0 and 1)\nnum=5:  bits |= bits<<5 → can reach 0,1,5,6\nnum=11: bits |= bits<<11 → can reach 0,1,5,6,11,12,...\nBit 11 is set → return true\n✅ Answer: true`,
      keyInsight: `O(n × target / 64) time effectively — bitwise parallelism means up to 64 sum values are processed per CPU instruction. Space is O(target / 64). The fastest practical approach for large targets.`,
    },
  },
}
